plugins/woocommerce-blocks/assets/js/base/components/combobox/index.tsx

```diff
@@ -88,72 +124,137 @@ const Combobox = ( {
 		setValidationErrors,
 	] );
 
-	// @todo Remove patch for ComboboxControl once https://github.com/WordPress/gutenberg/pull/33928 is released
-	// Also see https://github.com/WordPress/gutenberg/pull/34090
+	const outerWrapperClasses = classnames(
+		'wc-block-components-combobox',
+		className || '',
+		{
+			'is-active': selectedOption?.value,
+			'has-error': error?.message && ! error?.hidden,
+		}
+	);
+
+	const ariaInvalid = error?.message && ! error?.hidden ? 'true' : 'false';
+	const store = useComboboxStore();
+	const activeValue = store.useState( 'activeValue' );
+
+	const onClose = useCallback( () => {
+		if ( searchTerm.length ) {
+			const bestMatch = findBestMatchByLabel( searchTerm, options );
+
+			if ( bestMatch ) {
+				setSearchTerm( bestMatch.label );
+				setSelectedOption( bestMatch );
+				onChange( bestMatch.value );
+			} else {
+				setSearchTerm( selectedOption?.label || '' );
+			}
+			// Search term was cleared, reselect the last selected option
+		} else {
+			setSearchTerm( selectedOption?.label || '' );
+		}
+	}, [
+		searchTerm,
+		options,
+		onChange,
+		setSelectedOption,
+		setSearchTerm,
+		selectedOption,
+	] );
+
 	return (
 		<div
-			id={ controlId }
-			className={ classnames( 'wc-block-components-combobox', className, {
-				'is-active': value,
-				'has-error': error?.message && ! error?.hidden,
-			} ) }
 			ref={ controlRef }
+			id={ controlId }
+			className={ outerWrapperClasses }
 		>
-			<ComboboxControl
-				className={ 'wc-block-components-combobox-control' }
-				label={ label }
-				onChange={ onChange }
-				onFilterValueChange={ ( filterValue: string ) => {
-					if ( filterValue.length ) {
-						// If we have a value and the combobox is not focussed, this could be from browser autofill.
-						const activeElement = isObject( controlRef.current )
-							? controlRef.current.ownerDocument.activeElement
-							: undefined;
-
-						if (
-							activeElement &&
-							isObject( controlRef.current ) &&
-							controlRef.current.contains( activeElement )
-						) {
-							return;
-						}
+			<div className="components-combobox-control components-base-control wc-block-components-combobox-control">
+				<ComboboxProvider
+					store={ store }
+					value={ searchTerm }
+					selectedValue={ selectedOption?.value || '' }
+					setValue={ ( val: string ) => {
+						startTransition( () => {
+							setSearchTerm( val );
 
-						// Try to match.
-						const normalizedFilterValue =
-							filterValue.toLocaleUpperCase();
+							if ( val && val.length ) {
```

Comment:
Nitpic: What are your thoughts on writing this conditional a bit more concise? 

```suggestion
							if ( val?.length ) {
```